src/Plugins/RestServer/Tokens/NEP11Token.cs

```diff
@@ -0,0 +1,179 @@
+            ArgumentNullException.ThrowIfNull(scriptHash, nameof(scriptHash));
+            _neoSystem = neoSystem;
+            _snapshot = snapshot ?? _neoSystem.GetSnapshotCache();
+            _contract = NativeContract.ContractManagement.GetContract(_snapshot, scriptHash) ?? throw new ArgumentException(null, nameof(scriptHash));
+            if (ContractHelper.IsNep11Supported(_contract) == false)
+                throw new NotSupportedException(nameof(scriptHash));
+            Name = _contract.Manifest.Name;
+            ScriptHash = scriptHash;
+
+            byte[] scriptBytes;
+            using var sb = new ScriptBuilder();
+            sb.EmitDynamicCall(_contract.Hash, "decimals", CallFlags.ReadOnly);
+            sb.EmitDynamicCall(_contract.Hash, "symbol", CallFlags.ReadOnly);
+            scriptBytes = sb.ToArray();
+
+            using var appEngine = ApplicationEngine.Run(scriptBytes, _snapshot, settings: _neoSystem.Settings, gas: RestServerSettings.Current.MaxGasInvoke);
+            if (appEngine.State != VMState.HALT)
+                throw new NotSupportedException(nameof(ScriptHash));
+
+            Symbol = appEngine.ResultStack.Pop().GetString() ?? throw new ArgumentNullException();
+            Decimals = (byte)appEngine.ResultStack.Pop().GetInteger();
+        }
+
+        public BigDecimal TotalSupply()
+        {
+            if (ContractHelper.GetContractMethod(_snapshot, ScriptHash, "totalSupply", 0) == null)
+                throw new NotSupportedException(nameof(ScriptHash));
+            if (ScriptHelper.InvokeMethod(_neoSystem.Settings, _snapshot, ScriptHash, "totalSupply", out var results))
+                return new(results[0].GetInteger(), Decimals);
+            return new(BigInteger.Zero, Decimals);
+        }
+
+        public BigDecimal BalanceOf(UInt160 owner)
+        {
+            if (ContractHelper.GetContractMethod(_snapshot, ScriptHash, "balanceOf", 1) == null)
+                throw new NotSupportedException(nameof(ScriptHash));
+            if (ScriptHelper.InvokeMethod(_neoSystem.Settings, _snapshot, ScriptHash, "balanceOf", out var results, owner))
+                return new(results[0].GetInteger(), Decimals);
+            return new(BigInteger.Zero, Decimals);
+        }
+
+        public BigDecimal BalanceOf(UInt160 owner, byte[] tokenId)
+        {
+            if (Decimals == 0)
+                throw new InvalidOperationException();
+            if (ContractHelper.GetContractMethod(_snapshot, ScriptHash, "balanceOf", 2) == null)
+                throw new NotSupportedException(nameof(ScriptHash));
+            ArgumentNullException.ThrowIfNull(tokenId, nameof(tokenId));
+            if (tokenId.Length > 64)
+                throw new ArgumentOutOfRangeException(nameof(tokenId));
+            if (ScriptHelper.InvokeMethod(_neoSystem.Settings, _snapshot, ScriptHash, "balanceOf", out var results, owner, tokenId))
+                return new(results[0].GetInteger(), Decimals);
```

Comment:
If this conversion throw an exception, It will be catched by a middleware, isn't it?